ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+building concurrent applications (e.g. Async or EventMachine). These libraries allow users to
+concurrently execute IO operations.
+
+There are however quite a few big limitations to this approach:
+
+- You need to use very specific libraries to perform IO calls this way (there isn't one for gRPC)
+- Any CPU-bound instructions block the reactor and therefore need to be executed in small bursts per
+  reactor tick
+- Any user-defined code can block the whole thing and has to be wrapped in threads anyways
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, options: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
+  url: 'localhost:7233'
+)
+
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'Alice', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+
+worker = Temporal::Worker.new(config, 'my-namespace', 'my-queue')
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
+- Configuration will provide defaults where possible (avoiding extensive upfront configuration)
+- Configuration will be used for other more advanced options such as logger, metrics, payload
+  converters, middleware, etc
+
+```ruby
+class Temporal::Client
+  # Creates a new instance of a Client
+  def self.new(config: Temporal::Configuration, **options) -> Temporal::Client
+
+  # Start the workflow and return handle for further interaction
+  # TODO: expand on :options
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
```

Comment:
I don't believe you can do this in a reasonable way and be cross-language compatible. In Python, we do not support kwargs for workflows/activities for many reasons. The most obvious of which is we have no way to track names. There are other reasons like compatibility expectation on name change, that we suggest a single class for more than one param, and that we need to be able to invoke Ruby workflows from non-Ruby SDKs. Also, how does this work if I have a workflow parameter named options?

Also, when you don't allow dynamic kwargs, you get added benefits such as using those kwargs yourself. I know this takes an options argument, but if it were idiomatic on some calls to instead just take different kwargs for options, you could.

Reply:
Hmm… I don't know how kwargs work in Python, but in Ruby kwargs can be defined after positional args only (if there are any) and they aren't too different from a Hash (dictionary) with Symbols used as keys. Basically:

```ruby
def test(a, b:)
  ...
end

# When called directly
test(a, b: 2)

# When called without knowing exact args/kwargs
args = [1]
kwargs = { b: 2 }
test(*args, **kwargs)


def test_2(*args, **kwargs)
  ...
end

# This is the reverse case, when we accept any args and kwargs
test_2(1, b: 2)
```

This means that we can wrap all kwargs in something and treat them it in a special way. I don't think cross-language compatibility here would be an issue, but I'm just not sure how these are handled in Python. If it's a special kind of payload that kwarg-supporting languages can parse (and all others — assemble) then I don't think it should be a problem.

I think not allowing these in Ruby if other languages can't use them is a weird choice. I think cross-language use case is probably more rare than a single-language. So which one do we optimise for?

You're right about `options`, these are taken. I've renamed these to `workflow:` to reduce the chance of a collision.

Reply:
> I don't know how kwargs work in Python, but in Ruby kwargs can be defined after positional args only

They work the same way in Python

> If it's a special kind of payload that kwarg-supporting languages can parse

I think we really want to avoid special kinds of payloads. So much so we have disallowed kwargs in Python workflows/activities even though they are _very_ common in the language.

> I think not allowing these in Ruby if other languages can't use them is a weird choice. I think cross-language use case is probably more rare than a single-language. So which one do we optimise for?

There are a lot of constraints we place on functions. The key is to optimize for consistency.

Also, we encourage people with multiple parameters to use a class in these cases. We don't want people using multiple arguments because it is not very future proof (but we support them because we always have). Adding kwargs is even worse. In Python we don't allow default parameter values either. People really need to be using language-native structures if they need more than one param and we should really encourage that, not provide workarounds to the things we discourage.

Reply:
Not sure I get the reasoning behind this:

1. We want to disallow kwargs because they result in special payload types and won't get cross-language support
2. We encourage people to use custom classes to pass multiple arguments together

But 2 is implying that you need a special type of payload and will have an even harder time making it work across languages. On top of that you lose static type safety for the specific params.

Am I missing something?

Reply:
Discussed off-PR, but most languages have preferred approaches to serializing/deserializing JSON objects. Our default serialization mechanism is JSON. Now, how to discover the types to deserialize back into instead of a hash map or whatever Ruby JSON does by default is the subject of a different discussion on type discoverability.